c10/core/InferenceMode.h

```diff
@@ -0,0 +1,67 @@
+#pragma once
+
+#include <c10/macros/Macros.h>
+#include <c10/core/impl/LocalDispatchKeySet.h>
+
+namespace c10 {
+
+// A RAII, thread local (!) guard that enables or disables inference mode upon
+// construction, and sets it back to the original value upon destruction.
+struct TORCH_API InferenceMode {
+  InferenceMode(bool enabled=true) : prev_mode(is_enabled()),
+   prev_keyset(c10::impl::tls_local_dispatch_key_set()) {
+    // Expected state:
+    //   InferenceMode: no InplaceOrView in included, Autograd in excluded
+    //   NormalMode: InplaceOrView in included, no Autograd in excluded
```

Comment:
An alternative that feels more symmetric to me is (1) put InplaceOrView on tensors (like Autograd), and (2) add InplaceOrView and Autograd both to excluded when you turn on InferenceMode. I wonder if there is some semantic problem with doing it this way.

Reply:
Is there ever a case where InplaceOrView is excluded but inference mode TLS is not true?

Reply:
> The proposed idea won't work since in this way both your "normal tensor" and "inference tensor" (tensors created in InferenceMode) will look the same (they both have Autograd and InplaceOrView keys).

Hmm, I don't follow. Currently the TensorImpl construction logic is:

```
 if (c10::InferenceMode::is_enabled()) {
    // See Note [Expected TLS state in InferenceMode] for why we don't add Autograd & InplaceOrView keys.
    key_set_ = key_set;
  } else {
    // TODO: Ideally we only add AutogradBackend key when the tensor requires grad.
    //       See Note [Dream: skip VariableType kernel when requires_grad=false]
    DispatchKey k = key_set.highestPriorityBackendTypeId();
    key_set_ = key_set | getAutogradRelatedKeySetFromBackend(k);
  }
```

It seems to me that you would want to *unconditionally* set InplaceOrView, but modulate Autograd on whether or inference mode is enabled. Then a normal tensor will have Autograd and InplaceOrView, but a inference mode tensor will only have InplaceOrView.

Reply:
> Currently no, InplaceOrView is only removed from included when inference mode TLS is true. Are you thinking about we can eliminate the inference mode TLS and check InplaceOrView key instead? ;)

Perhaps. It's a little indirect so it might not be good for understandability, but it certainly would reduce the amount of TLS you have to do.

Reply:
> inference mode tensor will only have InplaceOrView.

You don't actually want that. You want Inference Tensor to have that key only if they are a view of a non-inference mode Tensor.
All other inference mode Tensors should not have that key.

Reply:
Hmmm I'm happy to take that indirection if it reduces a TLS :D.